stats/statsd/common/collector.go

```diff
@@ -0,0 +1,205 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2016 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package common
+
+import (
+	"bytes"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/stats"
+	"github.com/loadimpact/k6/ui"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	pushInterval = 1 * time.Second
+)
+
+var _ lib.Collector = &Collector{}
+
+// Collector defines a collector struct
+type Collector struct {
+	Client *statsd.Client
+	Config Config
+	Logger *log.Entry
+	Type   ClientType
+	// FilterTags will filter tags and will return a list representation of them if it's not set
+	// tags are not being sent
+	FilterTags func(map[string]string, string) []string
+	Summary    map[string]*stats.Metric
+
+	startTime  time.Time
+	buffer     []*Sample
+	bufferLock sync.Mutex
+}
+
+// Init sets up the collector
+func (c *Collector) Init() error {
+	c.Summary = make(map[string]*stats.Metric)
+	return nil
+}
+
+// Link returns the address of the client
+func (c *Collector) Link() string {
+	return c.Config.Addr.String
+}
+
+// Run the collector
+func (c *Collector) Run(ctx context.Context) {
+	c.Logger.Debugf("%s: Running!", c.Type.String())
+	ticker := time.NewTicker(pushInterval)
+	c.startTime = time.Now()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.pushMetrics()
+		case <-ctx.Done():
+			c.pushMetrics()
+			c.finish()
+			return
+		}
+	}
+}
+
+// GetRequiredSystemTags Return the required system sample tags for the specific collector
+func (c *Collector) GetRequiredSystemTags() lib.TagSet {
+	return lib.TagSet{} // no tags are required
+}
+
+// SetRunStatus does nothing in statsd collector
+func (c *Collector) SetRunStatus(status lib.RunStatus) {}
+
+// Collect metrics
+func (c *Collector) Collect(containers []stats.SampleContainer) {
+	var pointSamples []*Sample
+
+	for _, container := range containers {
+		for _, sample := range container.GetSamples() {
+			c.Summary[sample.Metric.Name] = sample.Metric
```

Comment:
This seems like something that should be behind the `c.bufferLock`, unless we want a data race

Reply:
I removed it either way ... the whole way the summary was made was not what I would've wanted and if we actually want to send the summary I suppose we will just implement it centrally